### ISSUE TITLE
Unusable in Cloudflare workers due to use of node specific APIs/packages

### ISSUE THREAD

**Issue by cal-smith** [stripe/stripe-node#1593]
### Describe the bug

This seems to be a broad regression, given that Cloudflare workers is [supposed to be supported](https://blog.cloudflare.com/announcing-stripe-support-in-workers/).

The root of the issue is usage of node specific APIs and globals.
- webhooks uses Buffer - not supported in workers, one workaround would be to wrap it in a try/catch and fall back to assuming the value is a string (or I guess just check if it exists and fall back)
- utils requires `crypto` which is required early in stripe.ts - it's plausible that import could be replaced/polyfilled with Web Crypto
- there's a few references to `http`/`https` in `createNodeHttpClient`- the compiler doesn't seem to be smart enough to skip compiling NodeHttpClient (I think since it's imported by default? It's possible that dynamically requiring it could help?)
- somewhere there's `child_process` usage

I was able to get things mostly working by using the package.json `browser` field to stub out the `crypto`(swapped for `crypto-js`)/`http`/`https`/`child_process`(swapped the others for an empty dummy.ts file) imports, but webhooks unchecked Buffer usage fully blocked me.

### To Reproduce

I'll see if I can get a repro repo up but:
- create a new cloudflare function
- install stripe-node
- observe many errors

### Expected behavior

Everything should work!

### OS

Ubuntu

### Node version

19

### Library version

10.15.0

### API version

22-08-01

### Additional context

I'm @cal on slack if chatting synchronously helps!

**Comment by kamil-stripe** [comment]
Hi @cal-smith! Thank you for the report. Are you able to share your `wrangler.toml` with us?
We are trying to make our stripe-node less Node.js-dependent but I just verified it still works with `node_compat = true` setting in your configuration. Please take a look at our [sample CloudFlare Worker project](https://github.com/stripe-samples/stripe-node-cloudflare-worker-template) too.

**Comment by kamil-stripe** [comment]
I see! I don't believe we have tested our library with that environment. Let me investigate how can we support it and get back to you.

**Comment by pakrym-stripe** [comment]
I apologize for the delayed response @cal-smith.

Indeed, [CloudFlare Functions](https://developers.cloudflare.com/pages/platform/functions/) do not allow the same level of configuration as workers.

I'll check with the team to see if we can do something short-term for Buffer usages and/or long-term to make the stripe SDK work with workers/functions natively.

**Comment by pakrym-stripe** [comment]
Just discovered that passing `--node-compat` is allowed in `wrangler pages dev`:

```
wrangler pages dev . --node-compat
```

**Comment by CanRau** [comment]
just ran into the same issue, unfortunately `--node-compat` seems not (yet) supported in Pages according to https://github.com/cloudflare/wrangler2/issues/1074 & the mentioned & closed PR https://github.com/cloudflare/wrangler2/pull/1857

**Comment by colinhowells** [comment]
Been struggling with this for a while; I'm dealing with this by separating all Stripe things from my Pages project into a Worker which the Pages project calls as an external api so I can use `node_compat` on it.

Based on forks I've seen[0] of the `stripe` lib  I _think_ creating a non-Node version of it shouldn't be too difficult – solving the problem (admittedly Cloudflare's problem by announcing something that wasn't true in all use cases) by resorting to a fork shouldn't even be a temptation given how secure payments need to be.

[0] `https://www.npmjs.com/package/stripe-cloudflare-worker`

**Comment by CanRau** [comment]
yea def not gonna use a non-official fork, tho using a Worker in addition is interesting, but I think I switch our SvelteKit to use the Workers adapter instead and get logging as well 🤔 had a custom GitHub Actions pipeline planned anyway 💁🏼

**Comment by CanRau** [comment]
As the Worker logged the same errors I switched to using Stripe's REST API instead, works great so far.

For response types I just use the ones from `stripe` 🤓

**Comment by colinhowells** [comment]
I think we're both doing the same thing then – making Stripe calls in a Worker we can run `node_compat` on?

(`stripe` is a [rest api client](https://stripe.com/docs/api) and needs to be server-side to avoid key exposure)

If you have something that sketches out your approach I'm curious. Creating a separate Worker to serve as a Stripe api my Pages project calls is a good workaround so far, until Pages reaches parity with Workers or Stripe offers a [WinterCG](https://wintercg.org/) library

**Comment by CanRau** [comment]
no separate worker, I'm have helper functions which directly call the rest api like so

```ts
const body = new URLSearchParams();
body.append("email", email);
body.append("name", name);
body.append("metadata[uid]", userId);

const response = await fetch("https://api.stripe.com/v1/customers", {
	method: "POST",
	body,
	headers: {
		Authorization: `Bearer ${STRIPE_SECRET_API_KEY}`,
		"Idempotency-Key": userId,
		"Content-Type": "application/x-www-form-urlencoded",
	},
});
```
they live in a `stripe.server.ts` file in our SvelteKit project so are only used and included in the server bundle

**Comment by colinhowells** [comment]
Ah, that's not a bad approach, that hadn't occurred to me. I'm using SvelteKit as well, I might try that out, thanks for that!

**Comment by CanRau** [comment]
Yea had to get inspiration from some other related issue comment, even though I'm doing the same with Mailjet 😅

So far the rest API is pretty straight forward if you have questions or anything let me know 🙌

**Comment by Cherry** [comment]
As another heads up @kamil-stripe, it might be worth updating your example Cloudflare Worker to Wrangler 2. Wrangler 1 is deprecated. I've got a PR at https://github.com/stripe-samples/stripe-node-cloudflare-worker-template/pull/3.

**Comment by anniel-stripe** [comment]
Hi everyone, thanks for the wait! These compiler / runtime errors should be resolved with [v11.10.0](https://github.com/stripe/stripe-node/releases/tag/v11.10.0), where we now export a version of stripe-node for worker environments that does not depend on Node-specific APIs and globals.

If you encounter issues with the latest version, please feel free to reopen this issue or create a new one.
